BrowserView: Updated, rebased PR with bug fixes.

kanishk98 opened the pull request:
---
<!--
Remove the fields that are not appropriate
Please include:
-->

**What's this PR do?**

Same purpose as #793, but updated to pass tests and include more recent features. 
For code to pass compilation, we may have to modify `electron.d.ts` to the file [here](https://gist.github.com/kanishk98/866ebbb3f42344165f3ec6f8b8750029).

I needed to make the above changes so that `page-title-updated` would be counted as a `before-input-event`.

**You have tested this PR on:**
  - [x] Windows
  - [ ] Linux/Ubuntu
  - [ ] macOS

vsvipul committed:
BrowserView: Add View wrapper Class for BrowserView.

View extends the BrowserView class in the main
process and provides and interface similar to
webview. Other functions can be added later
when required.

vsvipul committed:
BrowserView: Add ViewManager for managing multiple Views.

Adds ViewManager class which makes it easier
to create, manage and delete views. Also,
created some basic listeners which will be
required later.

vsvipul committed:
BrowserView: Add BrowserView to index.ts.

vsvipul committed:
BrowserView: Remove webview props from tab.ts.

vsvipul committed:
BrowserView: Remove webview usage and initialize BrowserView.

Removes all instances of webview from main.ts and
adds an initial working instance of BrowserView.

vsvipul committed:
BrowserView: Add logic for badgeCount update.

Implements logic for updating badgeCount
to show message count in sidebar.

vsvipul committed:
BrowserView: Remove tooltips and use title instead.

Removes tooltips that we show for orgs and settings,
as these tooltips are no more visible over the
BrowserViews, so there is no use of keeping them.
Instead, uses title property of HTMLElements, which
can be displayed over BrowserViews.

vsvipul committed:
BrowserView: Add favicon-update event and fix view function calls.

vsvipul committed:
BrowserView: Destroy all views during reload-app.

Destroys all views to recreate them later
when app is reloaded. Also, set lastActiveIndex
to 0 when a tab is removed.

vsvipul committed:
BrowserView: Improvements to loading indicator.

vsvipul committed:
BrowserView: Improve ipc names.

vsvipul committed:
BrowserView: Add logic to handle external links.

vsvipul committed:
BrowserView: Rename canGoBackButton to maybeEnableGoBackButton.

vsvipul committed:
BrowserView: Implement userAgent and updateBadge onload.

Add logic to set user agent for each browserview.
Also, the badgeCount was not shown initially,
on opening the app. Adds that too.

vsvipul committed:
BrowserView: Add logic to show network error page.

vsvipul committed:
BrowserView: Add logic for custom CSS.

vsvipul committed:
BrowserView: Call fixBounds before setting view.

vsvipul committed:
BrowserView: Update deps.

vsvipul committed:
gitignore: Add .vscode to gitignore.

vsvipul committed:
BrowserView: Show views only when dom content loaded.

Adds a refresh function which refreshes view
every 200 ms, and shows it only if the view
DOM content is loaded. Also, gets rid of buggy
browser view switching which we were facing earlier.

vsvipul committed:
BrowserView: Rename webview instances to view and handle tray toggle.

vsvipul committed:
BrowserView: Add decoded server name when pushing to tabs.

vsvipul committed:
BrowserView: Make separate callViewFunction.

vsvipul committed:
BrowserView: Fix focus switching.

vsvipul committed:
Remove unnecessary translation-util.js file.

vsvipul committed:
BrowserView: Remove show dom content on load.

Removes the heavy setInterval which we were
using earlier to prevent heavy CPU Usage.
The main reason we were using a setInterval
earlier was because view switching was buggy.
This commit fixes that.

vsvipul committed:
BrowserView: Add documentation and fix fullScreen.

vsvipul committed:
BrowserView: Port reconnection logic to BV.

kanishk98 committed:
BrowserView: Make system-util runnable in both processes.

kanishk98 committed:
BrowserView: Get enabled flag for notifs settings.

kanishk98 committed:
BrowserView: Port notification settings to BV.

akashnimare commented:
@kanishk98 is this good to merge? If yes, then I'll need to cross-check the auto-update before merging.

kanishk98 commented:
Yeah, features and settings seem to be working pretty well for me. I'll do another check just to make sure, but I think it's good to be merged. 

Please also text me on czo if you need any help with testing the auto-update process.

akashnimare commented:
Okay, great. I'm testing the auto-update on macOS first. Can you do the same on Linux?

kanishk98 commented:
>  Can you do the same on Linux?

I can test it out on Windows (will let you know the results on czo). 
@muskankhedia would you mind testing this on Linux? I'll walk you through the process.

kanishk98 commented:
I've fixed the merge conflicts and also updated the broken link to the modified `electron.d.ts` file required. 
@akashnimare please have a look at this and let me know if it's good to go. 
@muskankhedia can you try running the auto-update on Windows and/or Linux and let us know how that goes? I'm getting the "Windows is unable to find Zulip.exe" error on my system, which is not related to the auto-update.

akashnimare commented:
Oh, that error is really frustrating. I'll check it out.

akashnimare commented on app/translations/en-GB.json:
```diff
@@ -53,6 +54,7 @@
 	"Organizations": "Organizations",
 	"Shortcuts": "Shortcuts",
 	"Settings": "Settings",
+<<<<<<< HEAD
```
@kanishk98 gotta update this file.

kanishk98 replied:
Updated. :)

vsvipul commented:
@akashnimare  @kanishk98 what's the current status on this ?

akashnimare commented:
@vsvipul hey, good to see you back :)

This is next in our roadmap. I'm releasing the stable release from the master, once done we can push a beta release with this PR.

akashnimare commented:
@manavmehta let's clean it up and merge.

manavmehta commented on app/renderer/js/components/tab.ts:
```diff
@@ -25,23 +21,19 @@ class Tab extends BaseComponent {
 		this.$el.addEventListener('mouseout', this.props.onHoverOut);
 	}
 
-	showNetworkError(): void {
-		this.webview.forceLoad();
-	}
```
@kanishk98 @akashnimare Do we need to totally remove this or need to create some alternative?
